src/terminal/parser/base64.cpp

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "base64.hpp"
+
+using namespace Microsoft::Console::VirtualTerminal;
+
+static const wchar_t base64Chars[] = L"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+static const wchar_t padChar = L'=';
+
+#pragma warning(disable : 26446 26447 26482 26485 26493 26494)
+
+// Routine Description:
+// - Encode a string using base64. When there are not enough characters
+//      for one quantum, paddings are added.
+// Arguments:
+// - src - String to base64 encode.
+// Return Value:
+// - the encoded string.
+std::wstring Base64::s_Encode(const std::wstring_view src) noexcept
+{
+    std::wstring dst;
+    wchar_t input[3];
+
+    auto iter = src.cbegin();
+    // Encode each three chars into one quantum (four chars).
+    while (iter < src.cend() - 2)
+    {
+        input[0] = *iter++;
+        input[1] = *iter++;
+        input[2] = *iter++;
+        dst.push_back(base64Chars[input[0] >> 2]);
+        dst.push_back(base64Chars[(input[0] & 0x03) << 4 | input[1] >> 4]);
+        dst.push_back(base64Chars[(input[1] & 0x0f) << 2 | input[2] >> 6]);
+        dst.push_back(base64Chars[(input[2] & 0x3f)]);
+    }
+
+    // Here only zero, or one, or two chars are left. We may need to add paddings.
+    if (iter < src.cend())
+    {
+        input[0] = *iter++;
+        dst.push_back(base64Chars[input[0] >> 2]);
+        if (iter < src.cend()) // Two chars left.
+        {
+            input[1] = *iter++;
+            dst.push_back(base64Chars[(input[0] & 0x03) << 4 | input[1] >> 4]);
+            dst.push_back(base64Chars[(input[1] & 0x0f) << 2]);
+        }
+        else // Only one char left.
+        {
+            dst.push_back(base64Chars[(input[0] & 0x03) << 4]);
+            dst.push_back(padChar);
+        }
+        dst.push_back(padChar);
+    }
+
+    return dst;
+}
+
+// Routine Description:
+// - Decode a base64 string. This requires the base64 string is properly padded.
+//      Otherwise, false will be returned.
+// Arguments:
+// - src - String to decode.
+// - dst - Destination to decode into.
+// Return Value:
+// - true if decoding successfully, otherwise false.
+bool Base64::s_Decode(const std::wstring_view src, std::wstring& dst) noexcept

```

Comment:
Should we reserve approx. enough space in `dst` so we don't have to keep growing it?

Reply:
Good point. I've added that. Thanks!